plugins/libplugin-pay.c

```diff
@@ -0,0 +1,780 @@
+		tlvstream_set_tu32(&cur->tlv_payload->fields, 4,
+				   p->start_block +
+				       p->route[hopcount - 1].delay);
+
+		if (root->payment_secret != NULL)
+			tlvstream_set_tlv_payload_data(
+			    &cur->tlv_payload->fields, root->payment_secret,
+			    root->amount.millisatoshis);
+	}
+
+	/* Now allow all the modifiers to mess with the payloads, before we
+	 * serialize via a call to createonion in the next step. */
+	payment_continue(p);
+}
+
+static void payment_sendonion(struct payment *p)
+{
+	struct out_req *req;
+	u8 *payload, *tlv;
+	req = jsonrpc_request_start(p->plugin, NULL, "createonion",
+				    payment_createonion_success,
+				    payment_rpc_failure, p);
+
+	json_array_start(req->js, "hops");
+	for (int i = 0; i < tal_count(p->createonion_request->hops); i++) {
+		json_object_start(req->js, NULL);
+		struct createonion_hop *hop = &p->createonion_request->hops[i];
+		json_add_node_id(req->js, "pubkey", &hop->pubkey);
+		if (hop->style == ROUTE_HOP_LEGACY) {
+			payload = tal_towire_legacy_payload(tmpctx, hop->legacy_payload);
+			json_add_hex_talarr(req->js, "payload", payload);
+		}else {
+			tlv = tal_arr(tmpctx, u8, 0);
+			towire_tlvstream_raw(&tlv, hop->tlv_payload->fields);
+			payload = tal_arr(tmpctx, u8, 0);
+			towire_bigsize(&payload, tal_bytelen(tlv));
+			towire(&payload, tlv, tal_bytelen(tlv));
+			json_add_hex_talarr(req->js, "payload", payload);
+			tal_free(tlv);
+		}
+		tal_free(payload);
+		json_object_end(req->js);
+	}
+	json_array_end(req->js);
+
+	json_add_hex_talarr(req->js, "assocdata",
+			    p->createonion_request->assocdata);
+
+	if (p->createonion_request->session_key)
+		json_add_secret(req->js, "sessionkey",
+				p->createonion_request->session_key);
+
+	send_outreq(p->plugin, req);
+}
+
+/* Mutual recursion. */
+static struct command_result *payment_finished(struct payment *p);
+
+/* A payment is finished if a) it is in a final state, of b) it's in a
+ * child-spawning state and all of its children are in a final state. */
+static bool payment_is_finished(struct payment *p)
+{
+	bool running_children = false;
+	if (p->step == PAYMENT_STEP_FAILED || p->step == PAYMENT_STEP_SUCCESS)
+		return true;
+	else if(p->step == PAYMENT_STEP_SPLIT || p->step == PAYMENT_STEP_RETRY) {
+		for (size_t i = 0; i < tal_count(p->children); i++)
+			running_children |= !payment_is_finished(p->children[i]);
+		return !running_children;
+	} else {
+		return false;
+	}
+}
+
+static enum payment_step payment_aggregate_states(struct payment *p)
+{
+	enum payment_step agg = p->step;
+
+	for (size_t i=0; i<tal_count(p->children); i++)
+		agg |= payment_aggregate_states(p->children[i]);
+
+	return agg;
+}
+
+/* A payment is finished if a) it is in a final state, of b) it's in a
+ * child-spawning state and all of its children are in a final state. */
+static bool payment_is_success(struct payment *p)
+{
+	return (payment_aggregate_states(p) & PAYMENT_STEP_SUCCESS) != 0;
```

Comment:
this seems like it'll return true if any of the child leaves are successful -- which may not indicate that all of its children are in a final state (as the code comment above indicates)

should you also check that the payment_is_finished in conjunction with the aggregate state check?

Reply:
The rationale behind this is that we want to show success as soon as possible, and not have to wait for some potentially stuck HTLCs for hours. In the non-MPP case the two considerations are identical, but for MPP we might have some in flight, while others are already settled with a preimage. I prefer showing success as soon as we have a preimage simply because it'll likely match what the recipient is seeing, i.e., the recipient has seen all pieces, and released the preimage, so from his point of view all is good.

This is also reflected in the lines below where we add a warning to that effect:

https://github.com/ElementsProject/lightning/blob/8bb9edc6dbb6c6906e93dbbf2568e40de391c8ea/plugins/libplugin-pay.c#L647-L652